### ISSUE TITLE
Speed up computed_summations

### ISSUE THREAD

**Issue by jakevdp** [PrincetonUniversity/FastTemplatePeriodogram#19]
This function is a big workhorse of the algorithm, and it could be sped up significantly, I think, using some vectorized operations rather than a triply-nested loop.

But first we need to unit test it. It might be best to first split it into two functions, one which computes the f-hat terms using the NFFT (which could be tested using a straightforward O[N^2] computation of the same thing) and one which uses these results to compute the C, S, YC, etc.

I was trying to understand this – are these sums just the standard Lomb-Scargle-type terms outlined in, e.g. the Zechmeister paper?

**Comment by johnh2o2** [comment]
Hi Jake,

That would be great -- these are "generalized" versions of the C, S, YC, ... values given in the Zechmeister & Kurster paper; CC, SS, and CS can be thought of as covariance matrices (e.g. CC_{nm} = cov(cos(nwt), cos(mwt))), while in the original K&M paper they are scalar-valued covariances (i.e. CC_{zechmeister & kurster} = CC_{fast templ. per}[1][1]). 

I'm in the process of writing up some testing code that does exactly what you mentioned (O(N^2) direct calculations). To test periodogram results, I'm also using a grid search over the phase shift parameter. I'll hopefully have an update a bit later tonight.

**Comment by jakevdp** [comment]
Awesome! By the way, I've been playing with this a bit, and I'm having a really hard time understanding pynfft. For example, I'd like to input a regular grid, compute the NFFT, and confirm that it gives the same output as a standard FFT. But I can't seem to make sense of the output.

Did you find any good resources on this when you were working on this part of the code?

**Comment by johnh2o2** [comment]
Hey Jake,

I've added unit tests for the modeler class and for the NFFT sums; everything works on my machine, and we'll see if TravisCI sees any problems.

The NFFT has some really great and thorough theory papers; I mostly consulted 

https://www-user.tu-chemnitz.de/~potts/nfft/guide/nfft3.pdf

for some quick references, but Stefan Kunis's dissertation is even more thorough (though I have found a sign error in the "fast gaussian gridding" section):

http://www.analysis.uni-osnabrueck.de/kunis/paper/KunisDiss.pdf

I can dig up more resources if you'd like. One possible source of confusion are the conventions used by `NFFT`/`pyNFFT` that differ from `FFTW`; for instance, the output of NFFT is `f_hat[n]` from `-N/2` to `N/2` instead of from `0` to `N`. It's critical, also, that the time values be scaled/shifted to `[-1/2, 1/2)`.

Let me know if that helps at all.
